diff --git a/grammar.js b/grammar.js
index ef9b500..20936e3 100644
--- a/grammar.js
+++ b/grammar.js
@@ -45,6 +45,9 @@ export default grammar({
     $._block_string_start,
     $._block_string_content,
     $._block_string_end,
+
+    $._double_quote_string_content,
+    $._single_quote_string_content,
   ],
 
   supertypes: ($) => [$.statement, $.expression, $.declaration, $.variable],
@@ -430,10 +433,10 @@ export default grammar({
       ),
 
     _doublequote_string_content: ($) =>
-      repeat1(choice(token.immediate(prec(1, /[^"\\]+/)), $.escape_sequence)),
+      repeat1(choice($._double_quote_string_content, $.escape_sequence)),
 
     _singlequote_string_content: ($) =>
-      repeat1(choice(token.immediate(prec(1, /[^'\\]+/)), $.escape_sequence)),
+      repeat1(choice($._single_quote_string_content, $.escape_sequence)),
 
     _block_string: ($) =>
       seq(
diff --git a/src/scanner.c b/src/scanner.c
index e257c2d..3fa3663 100644
--- a/src/scanner.c
+++ b/src/scanner.c
@@ -11,6 +11,9 @@ enum TokenType {
   BLOCK_STRING_START,
   BLOCK_STRING_CONTENT,
   BLOCK_STRING_END,
+
+  DOUBLE_QUOTE_STRING_CONTENT,
+  SINGLE_QUOTE_STRING_CONTENT,
 };
 
 static inline void consume(TSLexer *lexer) { lexer->advance(lexer, false); }
@@ -154,9 +157,30 @@ static bool scan_comment_content(Scanner *scanner, TSLexer *lexer) {
   return false;
 }
 
+static bool scan_quote_string_content(char ending_char, TSLexer *lexer) {
+  bool has_content = false;
+
+  while (lexer->lookahead != 0 && lexer->lookahead != ending_char && lexer->lookahead != '\\') {
+    has_content = true;
+    consume(lexer);
+  }
+
+  return has_content;
+}
+
 bool tree_sitter_lua_external_scanner_scan(void *payload, TSLexer *lexer, const bool *valid_symbols) {
   Scanner *scanner = (Scanner *)payload;
 
+  if (valid_symbols[DOUBLE_QUOTE_STRING_CONTENT] && scan_quote_string_content('"', lexer)) {
+    lexer->result_symbol = DOUBLE_QUOTE_STRING_CONTENT;
+    return true;
+  }
+
+  if (valid_symbols[SINGLE_QUOTE_STRING_CONTENT] && scan_quote_string_content('\'', lexer)) {
+    lexer->result_symbol = SINGLE_QUOTE_STRING_CONTENT;
+    return true;
+  }
+
   if (valid_symbols[BLOCK_STRING_END] && scan_block_end(scanner, lexer)) {
     reset_state(scanner);
     lexer->result_symbol = BLOCK_STRING_END;
diff --git a/test/corpus/comments.txt b/test/corpus/comments.txt
index ba51866..fb00976 100644
--- a/test/corpus/comments.txt
+++ b/test/corpus/comments.txt
@@ -53,6 +53,13 @@ print("level 2 block comment")
 print("not level 2 block comment")
 --]==]
 
+print "--[[hello]]world" -- long comment delimiters inside a string
+print "--[=["
+print "foo"
+--]=]
+print "bar"
+print '---[==['           -- the same case in a single-quoted string
+
 --------------------------------------------------------------------------------
 
 (chunk
@@ -98,5 +105,36 @@ print("not level 2 block comment")
     arguments: (arguments
       (string
         content: (string_content))))
+  (comment
+    content: (comment_content))
+  (function_call
+    name: (identifier)
+    arguments: (arguments
+      (string
+        content: (string_content))))
+  (comment
+    content: (comment_content))
+  (function_call
+    name: (identifier)
+    arguments: (arguments
+      (string
+        content: (string_content))))
+  (function_call
+    name: (identifier)
+    arguments: (arguments
+      (string
+        content: (string_content))))
+  (comment
+    content: (comment_content))
+  (function_call
+    name: (identifier)
+    arguments: (arguments
+      (string
+        content: (string_content))))
+  (function_call
+    name: (identifier)
+    arguments: (arguments
+      (string
+        content: (string_content))))
   (comment
     content: (comment_content)))
diff --git a/test/highlight/string.lua b/test/highlight/string.lua
index 24af4af..718a622 100644
--- a/test/highlight/string.lua
+++ b/test/highlight/string.lua
@@ -57,3 +57,24 @@ _ = [[x]]
 
 _ = [=[x]=]
 --     ^ string
+
+_ = "--[[hello]]world"
+--  ^ string
+--       ^ string
+
+_ = "--[=["
+--  ^ string
+--     ^ string
+
+_ = "foo"
+--  ^ string
+
+--]=]
+--^ comment
+
+_ = "bar"
+--  ^ string
+
+_ = '---[==['
+--  ^ string
+--      ^ string
